### ISSUE TITLE
`polkadot-omni-node`: Merge (back) `bencher`, `chain-spec` (and even `try-runtime`)

### ISSUE THREAD

**Issue by kianenigma** [paritytech/polkadot-sdk#5567]
as new subcommands, but this time all of these commands are fully independent of the runtime. 

Note that this means these tools should still be available as a standalone CLI as well. Re-merging them into the `polkadot-omni-node` is more like a syntactic sugar to make life easier. See https://github.com/paritytech/polkadot-sdk/pull/5269#discussion_r1711224570

**Comment by seemantaggarwal** [comment]
Just putting whatever I find here: 

Entry Point: [The Subcommand BuildSpec](https://github.com/paritytech/polkadot-sdk/blob/f340e07ffd7cff69e930d19563186ff7f0fed601/cumulus/polkadot-omni-node/lib/src/cli.rs#L75)

which leads us here to [the definition](https://github.com/paritytech/polkadot-sdk/blob/f340e07ffd7cff69e930d19563186ff7f0fed601/templates/solochain/node/src/command.rs#L54)

which essentially is just running the command that is used to generate a chain specification file

**Comment by seemantaggarwal** [comment]
Now Looking at [the benchmark subcommand](https://github.com/paritytech/polkadot-sdk/blob/f340e07ffd7cff69e930d19563186ff7f0fed601/cumulus/polkadot-omni-node/lib/src/cli.rs#L105C12-L105C34)

Leads to the [definition here](https://github.com/paritytech/polkadot-sdk/blob/f340e07ffd7cff69e930d19563186ff7f0fed601/cumulus/polkadot-omni-node/lib/src/command.rs#L182)

**Comment by seemantaggarwal** [comment]
what I see, that we want to do is: 
right now this is implemented within the omninode and what we want to do is kind of derive this directly from the binary itself instead of writing a custom here and…
for that we want to make the chain spec library and the benchmarker library and then import it in the omniode binary.

**Comment by seemantaggarwal** [comment]
Investigating further:
https://github.com/paritytech/polkadot-sdk/blob/master/substrate/bin/utils/chain-spec-builder/src/lib.rs 
so chain spec is already a library

Even benchmarking is a crate? https://crates.io/crates/frame-benchmarking

So I guess it's just a matter of using this crate in and making it a dependency for polka omni node in terms of benchmarking subcommand?

**Comment by seemantaggarwal** [comment]
Summary: 
found out that benchmark cmd is already tapping into that logic which is used by omninode bencher binary via the frame benchmarking cli crate

but when it comes to buildspec cmd we found out that it's using  chain spec command behind the scenes

**Comment by seemantaggarwal** [comment]
Further discoveries: 
Will focus mainly on the chain-spec-builder first. 


`runner.sync_run(|config| cmd.run(config.chain_spec, config.network))`

let's go the run command here, and see what we find:

https://github.com/paritytech/polkadot-sdk/blob/f340e07ffd7cff69e930d19563186ff7f0fed601/substrate/client/cli/src/commands/build_spec_cmd.rs#L57
here we go:
Let me explain, what this does:

Checks if there are no boot nodes in the chain specification and if self.disable_default_bootnode is false.
If self.disable_default_bootnode  true, it generates a default boot node:
Converts network_config.node_key into a key pair.
Extracts the public key and derives a peer_id (a unique identifier for the node in the P2P network).
Constructs a multi-address for the node: 127.0.0.1:30333.
Adds the newly created boot node to the chain

**Comment by seemantaggarwal** [comment]
Now let's go to the actual chainspec to see what we have there: 

here are some commands within the chainspec builder

**Comment by seemantaggarwal** [comment]
the interesting one is [create](https://github.com/paritytech/polkadot-sdk/blob/f340e07ffd7cff69e930d19563186ff7f0fed601/substrate/bin/utils/chain-spec-builder/src/lib.rs#L219)

let's dive deeper into this
The core purpose of this is to generate a chain spec JSON file, which defines the parameters for running a blockchain network. The functions together:

Extract the runtime code (WASM blob).
Construct a chain specification.
Apply presets, patches, or full genesis configurations.
Generate and output the final JSON representation.


If the command is Create, the program calls generate_chain_spec_for_runtime(&cmd).
The resulting chain spec JSON is written to a file


Next, let's breakdown at key function here

**Comment by seemantaggarwal** [comment]
`generate_chain_spec_for_runtime`:
Gets the runtime code (WASM blob).
Set token properties (e.g., symbol "UNIT", decimals 12).
Create a ChainSpec builder.
Apply configuration (presets, patches, full config, or default config).
If it's a parachain, add parachain properties.
Return the JSON representation.


`get_runtime_code`:
This function fetches the runtime WASM code needed for chain execution.
If code is available in the command, use it.
Otherwise, read the WASM runtime file from disk.
Returns: A byte array (Cow<'static, [u8]>) containing the runtime WASM blob.


`process_action`:
This function applies the correct genesis configuration to the chain spec.
The builder starts with basic metadata but modifies the genesis state based on the action type.
Applies NamedPreset, Patch, Full, or Default config.
Builds and finalizes the ChainSpec.

**Comment by seemantaggarwal** [comment]
Chain of thought: I want to modify this Create
to be able to accept a Json, and do what run does essentially. 

Let's see how we can do that.

**Comment by seemantaggarwal** [comment]
let's see the differences in the flow: 

`generate_chain_spec_for_runtime(cmd)` generates the chain spec JSON for parachains and then 
`process_action(cmd, code, builder)` builds the chain spec based on presets, patches, or configs.

whereas run in BuildSpec allows generating and outputting a chain spec JSON based on the configuration.
Instead of manually building the chain spec, BuildSpec directly calls sc_service::chain_ops::build_spec(&*spec, raw_output).


It enhances the flow by allowing the user to build and print the complete chain spec, including default bootnodes.
generate_chain_spec_for_runtime(cmd) generates a specific runtime-based JSON, while BuildSpec constructs a full chain spec.

**Comment by seemantaggarwal** [comment]
What I missed in the plan above: 
I will also need to pass the network_config, I think we can send a default value for it, or ask the user to send a custom value 

also, we can try converting JSON string into actual ChainSpec object to see if the json provided is valid, if not, we can throw an error, 

Then add the bootnodes, and then Reconvert modified ChainSpec to JSON

**Comment by seemantaggarwal** [comment]
And then, the last should be (from what I understand):

Modify the Subcommand Enum in omni-node to call: 
Create(ChainSpecBuilderCmd),

 update Create so it accepts and processes the same parameters by Modifying the ChainSpecBuilderCmd struct

**Comment by skunert** [comment]
Thank you for looking into this!

If there are now major flaws, we don't really want to modify the logic of the individual binaries. Improving them is for sure possible, but that is its own issue.

# build-spec
The investigation needed here revolves mostly around good integration in the CLI. The overall goal is to re-use the clap structs already exposed by these tools, for example [here](https://github.com/paritytech/polkadot-sdk/blob/1866c3b4673b66a62b1eb9c8c82f2cd827cbd388/substrate/bin/utils/chain-spec-builder/src/lib.rs#L38-L38) for `chain-spec-builder`.

One example of what this could look like is:
```
# Everything after `chain-spec` is just exposed chain-spec-builder CLI
polkadot-omni-node chain-spec create --runtime something.wasm named-preset development
```
The problem discussed recently is about the overlap. The `build-spec` command offers some functionality overlap. This can be used to export chain-specs that have been embedded into the node. One example of this:
```
polkadot build-spec --chain westend > westend.json
```
This is possible because `westend` chain-spec is integrated into the polkadot node. This is defined [here](https://github.com/paritytech/polkadot-sdk/blob/1866c3b4673b66a62b1eb9c8c82f2cd827cbd388/polkadot/cli/src/command.rs#L77-L77). Since the omni-node is built to operate with many different runtimes, no chain-specs are integrated by default, making the `build-spec` command not very useful. It can still convert external chain-specs via `--raw` command, but the same is possible for `chain-spec-builder`.

## Way forward
The first step would be to integrate all the chain-spec-builder subcommands (not just Create) into the omni-node. Some things I can imagine:
1. `polkadot-omni-node chain-spec-builder <options>`
2. `polkadot-omni-node utils chain-spec-builder <options>`

Approach 1. is the simplest but also "pollutes" the CLI namespace. We already have tons of commands, having more at the top level is kind of convenient but increases the burden when trying to understand from reading the help.

Approach 2. would group all the integrated binaries into a `utils` subcommand. If this was the pattern to use, we could have all the exposed binary CLIs grouped there.

From gut feeling I am leaning towards approach 2. But happy to hear opinions about this, its hard to strike a good balance here. We can also not remove `build-spec`, because familiar users might expect it and it is available on our other binaries.

# benchmark
> Even benchmarking is a crate? https://crates.io/crates/frame-benchmarking

What you found here is the frame-side of benchmarking, which is used in the runtime. The CLI that is used by `frame-omni-bencher` is located in [benchmarking-cli](https://github.com/paritytech/polkadot-sdk/blob/1866c3b4673b66a62b1eb9c8c82f2cd827cbd388/substrate/utils/frame/benchmarking-cli/src/lib.rs#L20-L20).

If you look at `frame-omni-bencher`, it currently only supports the `benchmark pallet` and `benchmark overhead` subcommands: https://github.com/paritytech/polkadot-sdk/blob/1866c3b4673b66a62b1eb9c8c82f2cd827cbd388/substrate/utils/frame/omni-bencher/src/command.rs#L131-L149

`polkadot-omni-node` already supports a superset of this and uses `benchmarking-cli` behind the scenes too. So in effect `frame-omni-bencher v1 benchmark` and `polkadot-omni-node benchmark` are already the same.

The thing that is to be done here is to check that `polkadot-omni-node` is able to handle WASM blobs, I remember this being an issue in the past. So the task there is to verify that these commands work properly:

```
polkadot-omni-node benchmark pallet -h
...
      --runtime <RUNTIME>                                Optional runtime blob to use instead of the one from the genesis config
      --genesis-builder <GENESIS_BUILDER>                How to construct the genesis state [possible values: none, runtime, spec-runtime, spec-genesis]
      --genesis-builder-preset <GENESIS_BUILDER_PRESET>  The preset that we expect to find in the GenesisBuilder runtime API [default: development]
...
```

TLDR:
- Integrate chain-spec-builder CLI unaltered. Question: utils or top-level?
- Double check usage of `.wasm` files with `benchmark` command of `polkadot-parachain`

Also happy to hear @michalkucharczyk opinion on this. Since he has thought about CLI usability for some time.

**Comment by iulianbarbu** [comment]
### re: chain-spec-builder

Just wanted to mention that me and @seemantaggarwal  missed to clarify in our discussion that we'd want chain spec builder in full to be acceessible via 'polkadot-omni-node' (we focused only on how to fix the overlap in code, which looks right now more as a non-problem, since it doesn't block the actual goal of this issue, meaning integrating `chain-spec-builder` into omni node). Thanks for the reminder @skunert , and sorry @seemantaggarwal - I think I gave misleading direction, but still, it was useful context for both :).

> Since the omni-node is built to operate with many different runtimes, no chain-specs are integrated by default, making the build-spec command not very useful. It can still convert external chain-specs via --raw command, but the same is possible for chain-spec-builder.

That's the part which I think requires to be shared between the two. In general, build spec cmd for omni node is just setting bootnodes when they are passed via CLI & converts the chain spec to raw (again, if the raw flag is passed). My idea was to enable `chain-spec-builder` to patch an existing chain spec json file (currently it works only with runtimes as wasm blobs), and then `sc_cli::BuildSpec` functionality, when applied to omni node, could be easily replaced with `chain-spec-builder` API logic. Down the line, `chain-spec-builder` could provide API to operate over embedded  chain specs (like it is needed for `polkadot`), and `sc_cli` could have a direct dependency on it, making it shared logic as `frame-benchmarking-cli` is for `benchmark` cmd. However, the goal of this issue (integrating `chain-spec-builder` and `frame-omni-bencher` into `polkadot-omni-node` isn't depending directly on having `sc_cli::BuildSpecCmd` depend on `chain-spec-builder` logic, so we can leave that for later.

I am also fine with `polkadot-omni-node utils chain-spec-builder <options>`.

**Comment by kianenigma** [comment]
For chain spec builder

> I am also fine with polkadot-omni-node utils chain-spec-builder <options>.

also sounds good to me. Please adjust the CLI docs to highlight what the difference between this and `build-spec` then is? 

For benchmarking, so long as you can do your benchmarking in a e.g. template that has only a runtime with omni-node, we are good 🚀 

Then follow-ups would also be: 

1. adjust the README of all templates to take this into account, and not invite you to download also `chain-spec-builder` etc 
2. Make an issue in the `polkadot-docs` repo for the documentation team to update the zero-to-hero tutorial to this, and in a few months once this is released, see it through that it happens. @seemantaggarwal

**Comment by skunert** [comment]
Today we discussed this issue again, going in a different direction.

We discussed to not use the `utils` namespace, since it will be pretty empty. My original assumption was that we would have more tools there in the future, but for benchmarking for example it does not make sense.

So what we should do is:
- Add `chain-spec-builder` subcommand to the top-level namespace and expose its CLI there.
- Deprecate `build-spec` and introduce a dedicated command `export-chain-spec` to better reflect what is left of the command. Deprecation should be a good step since the redundancy in functionality is not helping anyone and the functionality should be reflected in the name.
- We also discussed what the additional parameters in `build-spec` are doing (e.g. `--base-path` and `--node-key`). I took a quick look and indeed they are not doing anything and are just inherited from the `SharedParams`. So these do not need to be preserved.

**Comment by seemantaggarwal** [comment]
PR: https://github.com/paritytech/polkadot-sdk/pull/7619

**Comment by seemantaggarwal** [comment]
From the discussion today, regarding [point number 2 which mentions deprecating build-spec](https://github.com/paritytech/polkadot-sdk/issues/5567#issuecomment-2664062515) : 
1. Idea is introduce a function export-chain-spec that does the chain-spec exporting. build-spec will work as before, but have a deprecation message on usage. 
2. We want to implement this subcommand within the omni-node binary only, to limit the scope here
3. For now, this new command will focus on the ability to take the chain-specs embedded in the node and "exports" them to a json file
4. We will not include the addition of boot node functionality, that the build-spec command has right now into the new command, if along the process, we decide otherwise, we can include it
5. Part of this PR will also be to display the message the build-spec will soon be on deprecation path 


Tagging; @skunert  to review and verify this plan

**Comment by iulianbarbu** [comment]
Opened #8594 to enable running `polkadot-omni-node benchmark pallet` without the `--chain` flag, since the subcommand can also use `--runtime` <wasm blob path> too. Apart from this, I noticed a few things in relation to benchmarking cli, and its support in various places:

* omni node doesn't support `overhead` and `extrinsics`
* frame-omni-bencher supports only `pallet` and `overhead`

Doesn't seem to be urgent to enable full feature parity for them, and we can also track it separately. I think it makes sense to have same subcommands in both `frame-omni-bencher` and `polkadot-omni-node` down the line, to not confuse people with what each binary provides. In terms of other nodes, it is questionable - if some node doesn't want to support a certain benchmarking-cli subcommand, then it should be able to optout.

Currently all commands show up in `help` for both binaries, only that a subset is supported (unsupported ones return early with a message clarifying the subcommand is not supported). I think this is fine as long as we show only the supported commands but at the same time I am not sure if we really want to opt out from supporting a certain benchmark command. To be discussed further. Will open an issue for this subject.

**Comment by iulianbarbu** [comment]
The only aspect left on this issue is integrating `try-runtime-cli` in omni-node, which for now can not be done easily because try-runtime-cli doesn't have a maintainer and is not in sync with `polkadot-sdk`. I extracted this subject here: https://github.com/paritytech/polkadot-sdk/issues/8762. I will close this issue until further notice.
